### ISSUE TITLE
[data grid] Add ability to wrap the row with contexts

### ISSUE THREAD

**Issue by deviprsd** [mui/mui-x#4739]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

Wrap the row with a custom `context` or `contexts` that can provide additional API to custom component cells. This can be achieved by `renderCell` but there are use cases where they are needed in a more row (parent) basis.

I also understand that this affects virtualization props being passed greatly as mistakes can be made, and people may start to render components in `rowContextWrapper` example below, so maybe passing the `provider` and `providerProps` to the data grid would be another option to prevent such concern.

### Examples 🌈

```tsx
rowContextWrapper: (row) => {
    return (
       <FormProvider {props}>
           {row}
       </FormProvider>
    )
}
```

### Motivation 🔦

The row can render cells with custom components. There are use cases where these custom components could benefit from additional APIs provided from the parent object, in this case the **row**.

One such concrete being able have custom input fields working with react-hook-form. It would be ideal to wrap each row with a react-hook-form `FormProvider` context for instance. This provides the ability for the custom components to edit the row data that may or may not affect other data points in that row.

**Comment by DanailH** [comment]
Hi, @deviprsd thanks for raising this. Wouldn't the `Row` component slot help you achieve this https://mui.com/x/api/data-grid/data-grid/#slots?

**Comment by deviprsd** [comment]
@DanailH Yes, that should work. I have been trying to find the docs for that. Part of the reason is it is not properly elaborated in the right side nav toc. The Props section could benefit from subsections

- Props
   - Slots

or slots should be their own section

- Props
- Slots

![Screen Shot 2022-05-04 at 9 46 52 AM](https://user-images.githubusercontent.com/3809354/166708221-73c7a0ef-cb8d-49fb-910d-465c6337fe53.png)

Each of these slots don't seem to have their own api reference anywhere in the docs as well, trying to find the information is slightly more difficult.

![Screen Shot 2022-05-04 at 9 57 37 AM](https://user-images.githubusercontent.com/3809354/166709373-a1586f1c-1ddc-4047-8665-451da556e555.png)
